packages/airnode-adapter/src/response-processing/casting.ts

```diff
@@ -53,26 +55,71 @@ function castBoolean(value: unknown): boolean {
   }
 }
 
-function castBytes32(value: any): string {
+function assertValueIsNotObject(value: unknown, targetType: 'bytes32' | 'string') {
   // Objects convert to "[object Object]" which isn't very useful
   if (isArray(value) || isPlainObject(value)) {
-    throw new Error(`Unable to convert: '${JSON.stringify(value)}' to bytes32`);
+    throw new Error(`Unable to convert: '${JSON.stringify(value)}' to ${targetType}`);
   }
+}
+
+function castString(value: any): string {
+  assertValueIsNotObject(value, 'string');
   return String(value);
 }
 
+function castBytes32(value: any): string {
+  assertValueIsNotObject(value, 'bytes32');
+
+  // NOTE: Do not use '.toString()' because it's not defined on 'null' value
+  const strValue = '' + value;
+
+  // We can't encode strings longer than 31 characters to bytes32.
+  // Ethers need to keep room for null termination
+  return strValue.length > 31 ? strValue.substring(0, 31) : strValue;
+}
+
+function castAddress(value: any): string {
+  // Objects convert to "[object Object]" which isn't very useful
+  if (isArray(value) || isPlainObject(value)) {
+    throw new Error(`Unable to convert: '${JSON.stringify(value)}' to address`);
+  }
+
+  const stringValue = String(value);
+  if (!ethers.utils.isAddress(stringValue)) {
+    throw new Error(`Unable to convert: '${stringValue}' to address`);
+  }
+  return stringValue;
+}
+
+function castHexString(value: any): string {
+  // Objects convert to "[object Object]" which isn't very useful
+  // also checks if the value is nil
+  if (isArray(value) || isPlainObject(value) || isNil(value)) {
+    throw new Error(`Unable to convert: '${JSON.stringify(value)}' to bytes`);
+  }
+  return ethers.utils.hexlify(value);
+}
+
 export function castValue(value: unknown, type: ResponseType): ValueType {
-  switch (type) {
-    case 'uint256':
-    case 'int256':
-      return castNumber(value, type);
+  if (isNumericType(type)) return castNumber(value, type);
+
+  const parsedArrayType = parseArrayType(type);
+  if (parsedArrayType) return applyToArrayRecursively(value, parsedArrayType, castNumber) as ValueType;
 
+  switch (type) {
     case 'bool':
       return castBoolean(value);
-
     case 'bytes32':
       return castBytes32(value);
+    case 'string':
+      return castString(value);
+    case 'address':
+      return castAddress(value);
+    case 'bytes':
+      return castHexString(value);
```

Comment:
Just out of curiosity, how will the requester decode a bytes data?

Will something like the following give us the expected result if we send the hexlified string?:
```
byte decodedData = abi.decode(data, (byte));
```

Reply:
🤔 that was my suspicion...maybe because the hex string starts with `"0x..."`? What I'm not sure about is if the hex string is the valid encoding for bytes.

Reply:
I mean that `byte` trigger compile error for me.

> What I'm not sure about is if the hex string is the valid encoding for bytes solidity type.

It does - I mean the e2e test is passing correctly. Or maybe I am misunderstanding something

Reply:
> I mean that `byte` trigger compile error for me.

It should be:
```
bytes decodedData = abi.decode(data, (bytes)); 
```
> It does - I mean the e2e test is passing correctly. Or maybe I am misunderstanding something

I'm just trying to make sure that `data` can be successfully decoded in the requester contract when type is `bytes`

Reply:
Yeah, but it does :D 

Or am I still misunderstanding?